### PR TITLE
25 get connection state

### LINKED ISSUE #25: Function to get state of a connection
Write a function that uses the NM DBus API to get the state a previously configured connection.
It should take a connection, it should return a result.

### PR DESCRIPTION
**Pull request by josephroberts**
Fixes #25

### CODE REVIEW

File: src/connection/mod.rs [hunk 1]
```diff
@@ -116,9 +119,54 @@ pub fn disable(c: &Connection, t: i32) -> Result<ConnectionState, String> {
...
+pub fn state(connection: &Connection) -> Result<ConnectionState, String> {
+    let active_paths = dbus_property!(NM_SERVICE,
+                                      NM_SERVICE_PATH,
+                                      NM_SERVICE_INTERFACE,
+                                      "ActiveConnections")
+        .inner::<&Vec<dbus::MessageItem>>()
+        .unwrap()
+        .iter()
+        .map(|p| extract_path(p.inner::<&dbus::Path>().unwrap().to_owned()))
+        .collect::<Vec<_>>();
```
Comment (zozo123) [inline src/connection/mod.rs]: tough one.

File: src/connection/mod.rs [hunk 2]
```diff
@@ -116,9 +119,54 @@ pub fn disable(c: &Connection, t: i32) -> Result<ConnectionState, String> {
...
+    // How can we avoid this ^ collect - it gets turned back into an iterator below.
+
+    let settings_paths = active_paths.iter()
```
Comment (zozo123) [inline src/connection/mod.rs]: Can we add one line of documentation what we do here?

File: src/connection/mod.rs [hunk 3]
```diff
@@ -116,9 +119,54 @@ pub fn disable(c: &Connection, t: i32) -> Result<ConnectionState, String> {
...
+        .map(|p| {
+            extract_path(dbus_property!(NM_SERVICE, p, NM_ACTIVE_INTERFACE, "Connection")
+                .inner::<&dbus::Path>()
+                .unwrap()
+                .to_owned())
+        });
+
+    let mut state = ConnectionState::Deactivated;
```
Comment (zozo123) [inline src/connection/mod.rs]: Please consider to add one line of description about the for loop logic.

File: src/connection/mod.rs [hunk 4]
```diff
@@ -155,6 +203,10 @@ fn get(path: dbus::Path) -> Result<Connection, String> {
     Ok(connection)
 }
 
+fn extract_path(path: dbus::Path) -> String {
+    path.as_cstr().to_str().unwrap().to_string()
```
Comment (josephroberts) [inline src/connection/mod.rs]: Done

**Pull request merged**